waspls/src/Wasp/LSP/Server.hs

```diff
@@ -0,0 +1,142 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Wasp.LSP.Server
+  ( run,
+  )
+where
+
+import qualified Control.Concurrent.MVar as MVar
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import qualified Control.Monad.Trans.Except as Except
+import qualified Control.Monad.Trans.State.Strict as State
+import qualified Data.Aeson as Aeson
+import Data.Default (Default (def))
+import qualified Data.Text as Text
+import qualified Language.LSP.Server as LSP
+import qualified Language.LSP.Types as J
+import System.Exit (ExitCode (ExitFailure), exitWith)
+import qualified System.Log.Logger
+import Wasp.LSP.Core (ServerConfig, ServerM, Severity (..), State)
+import Wasp.LSP.Handlers
+
+run :: Maybe FilePath -> IO ()
+run maybeLogFile = do
+  setupLspLogger maybeLogFile
+
+  state <- MVar.newMVar (def :: State)
+
+  let lspServerInterpretHandler env =
+        LSP.Iso {forward = runHandler, backward = liftIO}
+        where
+          runHandler :: ServerM a -> IO a
+          runHandler handler =
+            -- Get the state from the "MVar", run the handler in IO and update
+            -- the "MVar" state with the end state of the handler.
+            MVar.modifyMVar state \oldState -> do
+              LSP.runLspT env do
+                (e, newState) <- State.runStateT (Except.runExceptT handler) oldState
+                result <- case e of
+                  Left (severity, errMessage) -> sendErrorMessage severity errMessage
+                  Right a -> return a
+
+                return (newState, result)
+
+  exitCode <-
+    LSP.runServer
+      LSP.ServerDefinition
+        { defaultConfig = def :: ServerConfig,
+          onConfigurationChange = lspServerOnConfigChange,
+          doInitialize = lspServerDoInitialize,
+          staticHandlers = lspServerHandlers,
+          interpretHandler = lspServerInterpretHandler,
+          options = lspServerOptions
+        }
+
+  case exitCode of
+    0 -> return ()
+    n -> exitWith (ExitFailure n)
+
+-- | Setup DEBUG logger. Logs at other levels are ignored.
```

Comment:
Maybe just a short comment that explains this sets some global logger? Since that is somewhat unusual?

Reply:
Comment added